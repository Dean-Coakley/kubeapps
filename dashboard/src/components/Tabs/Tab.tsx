import * as React from "react";

export interface ITabProps {
  key: string;
  title: string;
  active: boolean;
  onClick: () => void;
}

class Tab extends React.Component<ITabProps> {
  public render() {
    const { title, active, onClick, key } = this.props;
    return (
      <div className={`Tabs__Tab ${active ? "Tabs__Tab-active" : ""}`}>
        <button
          type="button"
          onClick={onClick}
          role="tab"
          aria-controls={`${key}-content`}
          aria-selected={active}
        >
          {title}
        </button>
      </div>
    );
  }
}

export default Tab;
