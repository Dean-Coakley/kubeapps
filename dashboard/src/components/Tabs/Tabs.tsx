import * as React from "react";
import Tab from "./Tab";

export interface ITab {
  title: string;
  content: JSX.Element;
}

export interface ITabsProps {
  tabs: ITab[];
}

export interface ITabsState {
  tabActive: number;
}

class Tabs extends React.Component<ITabsProps, ITabsState> {
  public state: ITabsState = {
    tabActive: 0,
  };

  public render() {
    const { tabs } = this.props;
    const { tabActive } = this.state;
    return (
      <>
        <div className="Tabs" role="tablist">
          {tabs.map((tab, i) => {
            return (
              <Tab
                key={`tab-${i}`}
                title={tab.title}
                active={i === tabActive}
                onClick={this.activeTab(i)}
              />
            );
          })}
        </div>
        <div>
          {tabs.map((tab, i) => {
            const active = i === tabActive;
            const id = `tab-${i}-content`;
            return (
              <div key={id} id={id} role="tabpanel" hidden={!active} aria-expanded={active}>
                {tab.content}
              </div>
            );
          })}
        </div>
      </>
    );
  }

  private activeTab = (index: number) => {
    return () => {
      this.setState({ tabActive: index });
    };
  };
}

export default Tabs;
