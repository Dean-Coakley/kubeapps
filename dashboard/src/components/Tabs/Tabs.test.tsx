import { shallow } from "enzyme";
import * as React from "react";
import Tab from "./Tab";
import Tabs from "./Tabs";

const tabs = [{ title: "Foo", content: <div>foo</div> }, { title: "Bar", content: <div>bar</div> }];

it("renders a set of tabs", () => {
  const wrapper = shallow(<Tabs tabs={tabs} />);
  expect(wrapper).toMatchSnapshot();
});

it("shows by default the first tab", () => {
  const wrapper = shallow(<Tabs tabs={tabs} />);
  const tab = wrapper.find(Tab).filterWhere(t => t.prop("title") === tabs[0].title);
  expect(tab.prop("active")).toBe(true);
  const tabContent = wrapper.find("div").filterWhere(d => d.prop("id") === "tab-0-content");
  expect(tabContent.prop("hidden")).toBe(false);
});

it("shows the second tab content when clicked", () => {
  const wrapper = shallow(<Tabs tabs={tabs} />);
  let secondTab = wrapper.find(Tab).filterWhere(t => t.prop("title") === tabs[1].title);
  expect(secondTab.prop("active")).toBe(false);
  let tabContent = wrapper.find("div").filterWhere(d => d.prop("id") === "tab-1-content");
  expect(tabContent.prop("hidden")).toBe(true);

  secondTab.simulate("click");
  wrapper.update();

  secondTab = wrapper.find(Tab).filterWhere(t => t.prop("title") === tabs[1].title);
  tabContent = wrapper.find("div").filterWhere(d => d.prop("id") === "tab-1-content");
  expect(secondTab.prop("active")).toBe(true);
  expect(tabContent.prop("hidden")).toBe(false);
});
