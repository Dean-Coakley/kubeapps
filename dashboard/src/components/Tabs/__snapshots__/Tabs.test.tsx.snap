// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a set of tabs 1`] = `
<Fragment>
  <div
    className="Tabs"
    role="tablist"
  >
    <Tab
      active={true}
      key="tab-0"
      onClick={[Function]}
      title="Foo"
    />
    <Tab
      active={false}
      key="tab-1"
      onClick={[Function]}
      title="Bar"
    />
  </div>
  <div>
    <div
      aria-expanded={true}
      hidden={false}
      id="tab-0-content"
      key="tab-0-content"
      role="tabpanel"
    >
      <div>
        foo
      </div>
    </div>
    <div
      aria-expanded={false}
      hidden={true}
      id="tab-1-content"
      key="tab-1-content"
      role="tabpanel"
    >
      <div>
        bar
      </div>
    </div>
  </div>
</Fragment>
`;
